### PR TITLE
When i fork this repo i had some failures on branch develop

### PR DESCRIPTION
**Pull request by samuel20miglia**
Purpose of this pull request is to implement some tests, to make + stable + testable this already great framework.
When i fork this repo i had some failures on branch develop, so i have thinked to help.
` 
 public function testDispatchMiddlewareCanDispatchArrayOfMiddlewareAsMiddlewarePipe()
    {
        $middleware = [
            function () {
            },
            'FooBar',
            [InvokableMiddleware::class, 'staticallyCallableMiddleware'],
            InvokableMiddleware::class,
        ];

        $request = new Request([], [], '/', 'GET');
        $routeResult = RouteResult::fromRouteMatch(__METHOD__, $middleware, []);
        $request = $request->withAttribute(RouteResult::class, $routeResult);

        $container = $this->mockContainerInterface();
        $this->injectServiceInContainer($container, 'FooBar', function () {
        });

        $app = new Application($this->router->reveal(), $container->reveal());
        $app->dispatchMiddleware($request, new Response(), function () {
        });
    }

    public function testDispatchMiddlewareCanDispatchArrayOfMiddlewareAsMiddlewareExpressiveVersionTwoPipe()
    {
        $middleware = [
            function () {
            },
            'FooBar',
            [InvokableMiddleware::class, 'staticallyCallableMiddleware'],
            InvokableMiddleware::class,
            ];

        $middlewareNew = function ($req, $res, $next) {
        };

        $request = new Request([], [], '/', 'GET');
        $routeResult = RouteResult::fromRoute(new Route('/', $middlewareNew),$middleware);
        $request = $request->withAttribute(RouteResult::class, $routeResult);

        $container = $this->mockContainerInterface();
        $this->injectServiceInContainer($container, 'FooBar', function () {
        });

            $app = new Application($this->router->reveal(), $container->reveal());
            $app->dispatchMiddleware($request, new Response(), function () {
            });
    }

    public function uncallableMiddleware()
    {
        return [
            ['foo'],
            [['foo']]
        ];
    }

    /**
     * @dataProvider uncallableMiddleware
     * @expectedException \Zend\Expressive\Exception\InvalidMiddlewareException
     */
    public function testThrowsExceptionWhenDispatchingUncallableMiddleware($middleware)
    {
        $request = new Request([], [], '/', 'GET');
        $routeResult = RouteResult::fromRouteMatch(__METHOD__, $middleware, []);
        $request = $request->withAttribute(RouteResult::class, $routeResult);

        $this->getApp()->dispatchMiddleware($request, new Response(), function () {
        });
    }
`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,9 @@ zf-mkdoc-theme/
 clover.xml
 coveralls-upload.json
 zf-mkdoc-theme.tgz
+/.settings/
+/.buildpath
+/.project
+/composer.phar
+/deployment.properties
+/deployment.xml
```
Comment (xtreamwayz) [inline .gitignore]: These belong in .global_gitignore.

File: composer.lock [hunk 2]
```diff
@@ -4,7 +4,7 @@
         "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file",
         "This file is @generated automatically"
     ],
-    "content-hash": "8ecfa7d26c2ee0780ad83ff776fd8557",
+    "content-hash": "18b9c1c8d340ae62722961862eed3cca",
```
Comment (xtreamwayz) [inline composer.lock]: Why is composer.lock updated if you didn't change composer.json?

Comment (xtreamwayz) [inline composer.lock]: Oh, nvm. The lock file was out of date.

File: .gitignore [hunk 3]
```diff
@@ -5,3 +5,9 @@ zf-mkdoc-theme/
 clover.xml
 coveralls-upload.json
 zf-mkdoc-theme.tgz
+/.settings/
```
Comment (Ocramius) [inline .gitignore]: This entire file should be reverted

File: test/ApplicationTest.php [hunk 4]
```diff
@@ -240,6 +241,55 @@ public function testRouteAndDispatchMiddlewareAreNotPipedAtInstantation()
         $this->assertCount(0, $pipeline);
     }
 
+    public function testDispatchMiddlewareCanDispatchArrayOfMiddlewareAsMiddlewarePipe()
```
Comment (Ocramius) [inline test/ApplicationTest.php]: This test is missing assertions

File: test/ApplicationTest.php [hunk 5]
```diff
@@ -240,6 +241,55 @@ public function testRouteAndDispatchMiddlewareAreNotPipedAtInstantation()
...
+        $this->injectServiceInContainer($container, 'FooBar', function () {
+        });
+
+        $app = new Application($this->router->reveal(), $container->reveal());
+        $app->pipeDispatchMiddleware($request, new Response(), function () {
+        });
+    }
+
+
+    public function uncallableMiddleware()
```
Comment (Ocramius) [inline test/ApplicationTest.php]: Please document this provider

File: test/ApplicationTest.php [hunk 6]
```diff
@@ -240,6 +241,55 @@ public function testRouteAndDispatchMiddlewareAreNotPipedAtInstantation()
...
+    {
+        return [
+            ['foo'],
+            [['foo']]
+        ];
+    }
+
+    /**
+     * @dataProvider uncallableMiddleware
+     * @expectException Zend\Expressive\Exception\InvalidMiddlewareException
```
Comment (Ocramius) [inline test/ApplicationTest.php]: Please use `->expectException()` right before the `->pipeDispatchMiddleware()` call

**Comment by samuel20miglia** [comment]
ok thank you for support i'm working to fix all

File: composer.json [hunk 7]
```diff
@@ -1,79 +1,79 @@
...
-        ],
-        "upload-coverage": "coveralls -v",
-        "cs-check": "phpcs",
-        "cs-fix": "phpcbf",
-        "test": "phpunit --colors=always",
-        "test-coverage": "phpunit --coverage-clover clover.xml",
-        "license-check": "vendor/bin/docheader check src/ test/"
-    }
-}
+	"name": "zendframework/zend-expressive",
```
Comment (xtreamwayz) [inline composer.json]: Please use 4 spaces instead of a tab of 8. This goes for the entire file.

Comment (samuel20miglia) [inline composer.json]: ohh i'm sorry sure

**Comment by samuel20miglia** [comment]
i'm refactoring composer with 

> Please use 4 spaces instead of a tab of 8. This goes for the entire file.

File: .gitignore (hunk 3 above)
Comment (samuel20miglia) [inline .gitignore]: fixed

File: test/ApplicationTest.php (hunk 4 above)
Comment (tierresrl) [inline test/ApplicationTest.php]: fixed

File: .gitignore (hunk 1 above)
Comment (samuel20miglia) [inline .gitignore]: Fixed

File: test/ApplicationTest.php (hunk 5 above)
Comment (samuel20miglia) [inline test/ApplicationTest.php]: I have removed dataProvider  and implented test with comments

File: test/ApplicationTest.php (hunk 6 above)
Comment (samuel20miglia) [inline test/ApplicationTest.php]: fixed

**Comment by samuel20miglia** [comment]
I have fixed all reviews

File: composer.lock (hunk 2 above)
Comment (samuel20miglia) [inline composer.lock]: fixed

File: test/ApplicationTest.php (hunk 4 above)
Comment (samuel20miglia) [inline test/ApplicationTest.php]: Hi i have fixed this part.
Have a nice day

**Comment by caiquecastro** [comment]
The EOL were removed from some files.

File: composer.json [hunk 8]
```diff
@@ -1,79 +1,85 @@
...
+		],
+		"upload-coverage" : "coveralls -v",
+		"cs-check" : "phpcs",
+		"cs-fix" : "phpcbf",
+		"test" : "phpunit --colors=always",
+		"test-coverage" : "phpunit --coverage-clover clover.xml",
+		"license-check" : "vendor/bin/docheader check src/ test/"
+	},
+	"homepage" : "https://docs.zendframework.com/zend-expressive/",
+	"support" : {
```
Comment (xtreamwayz) [inline composer.json]: Move homepage and support to the top under description.

File: composer.json [hunk 9]
```diff
@@ -1,79 +1,85 @@
...
+		"issues" : "https://github.com/zendframework/zend-expressive/issues",
+		"source" : "https://github.com/zendframework/zend-expressive",
+		"irc" : "#zftalk.dev"
```
Comment (xtreamwayz) [inline composer.json]: You can remove irc as Slack is used.

File: composer.json [hunk 10]
```diff
@@ -1,79 +1,85 @@
...
-            "@test"
-        ],
-        "upload-coverage": "coveralls -v",
-        "cs-check": "phpcs",
-        "cs-fix": "phpcbf",
-        "test": "phpunit --colors=always",
-        "test-coverage": "phpunit --coverage-clover clover.xml",
-        "license-check": "vendor/bin/docheader check src/ test/"
-    }
+	"name" : "zendframework/zend-expressive",
```
Comment (xtreamwayz) [inline composer.json]: Why the tabs? What happened to the spaces?

File: composer.json [hunk 11]
```diff
@@ -1,79 +1,85 @@
...
+		"upload-coverage" : "coveralls -v",
+		"cs-check" : "phpcs",
+		"cs-fix" : "phpcbf",
+		"test" : "phpunit --colors=always",
+		"test-coverage" : "phpunit --coverage-clover clover.xml",
+		"license-check" : "vendor/bin/docheader check src/ test/"
+	},
+	"homepage" : "https://docs.zendframework.com/zend-expressive/",
+	"support" : {
+		"issues" : "https://github.com/zendframework/zend-expressive/issues",
```
Comment (froschdesign) [inline composer.json]: Please add also `forum` und `docs`. ([example](https://github.com/zendframework/zend-escaper/blob/master/composer.json#L10))

Comment (samuel20miglia) [inline composer.json]: ok thank you

File: composer.json (hunk 8 above)
Comment (samuel20miglia) [inline composer.json]: no problem, thank you

**Comment by froschdesign** [comment]
@samuel20miglia and @xtreamwayz 
I'm sorry, but I don't understand the entire goal of this PR.

* The title of this PR is too general.
* The description says nothing.
* Why is composer.json updated?
* Why different tests are updated?

Different problems are mixed here. This makes a review very difficult and no person can follow this PR.

**Comment by samuel20miglia** [comment]
I'm really sorry for this drawback,
i am php developer, my personal goal is contributing to a great platform as zf, zexpressive,
i use every day at work..., so help to make  + stable and + testable this already great framework.
@froschdesign  probably i have mixed to much, i'm sorry for this.

**Comment by froschdesign** [comment]
@samuel20miglia 
Thanks for your quick response and thanks for your work.

Can you edit the description of this PR and add a short explanation?

> so help to make + stable and + testable this already great framework.

👍

File: composer.json (hunk 8 above)
Comment (moderndeveloperllc) [inline composer.json]: I think that you have tabs at 8 spaces - Zend style uses 4 spaces per level in JSON.

**Comment by xtreamwayz** [comment]
@samsonasik 
> When i fork this repo i had some failures on branch develop, so i have thinked to help.

Can you explain what failures you exactly had? All I can see is you changed 1 docblock parameter and added some tests. No code is fixed or changed.

**Comment by samuel20miglia** [comment]
see on top please just edited this tests failed

File: composer.json (hunk 11 above)
Comment (samuel20miglia) [inline composer.json]: probably @froschdesign  suggest to close thi PR in front of open a new with clean code and goal.

File: composer.json (hunk 9 above)
Comment (samuel20miglia) [inline composer.json]: Fixed

File: composer.json (hunk 10 above)
Comment (samuel20miglia) [inline composer.json]: probably my zend studio....

**Pull request closed without merging**